### PR TITLE
Add tlsutil utility to get Determined Name from client's TLS certificate

### LINKED ISSUE #166: Unable to get info about original peer connecting to Gateway
Currently, the TLS support for GRPC and Gateway is only supporting encryption through those channels, without conveying any useful information about who is actually making the original connection.

This assumes that connections go as follows
1. From user/service .. as the `peer` to ...
2. Service `gateway` .. as the listening point .. to ..
3. The `grpc` layer, where any info is available in the `context.Context` on the implementation functions, due to gateway_proxy.go leveraging the following
  * setup ... `mux := runtime.NewServeMux(proxymeta.MetaOption(), runtime.WithIncomingHeaderMatcher(rpcutil.MatchHTTPHeaders))`
  * proxymeta.MetaOption() ServeMuxOption appears to add headers of interest defined in https://github.com/DecipherNow/gm-fabric-go/blob/master/metrics/headers/headers_of_interest.go to the map and involves something with routes
  * headers: `runtime.WithIncomingHeaderMatcher(rpcutil.MatchHTTPHeaders)` adds the http headers.

I think this issue can be addressed by analyzing the request and pulling the real peer info

# How I'm trying to get the peer info today from the context
```
func getPeerDN(w context.Context) string {
  p, pok := peer.FromContext(w)
  if pok && p.AuthInfo != nil {
    tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
    if ok && tlsInfo.State.HandshakeComplete {
      return tlsutil.GetDistinguishedName(tlsInfo.State.VerifiedChains[0][0])
    }
  }
  return ""
}
```

# Results Using TLS for GRPC + Gateway

settings.toml...
- grpc_use_tls = true
- gateway_use_tls = true
- use_gateway_proxy = true
- server_cert_path/key .. server.cert/key.pem for twl-server-generic2

1. When connecting as tester10,
   - Headers sent from client: none
   - AuthInfo has chain of
     - CN=twl-server-generic2,OU=DAE,OU=DIA,OU=twl-server-generic2,O=U.S. Government,C=US
     - CN=InterCA,OU=chimera,OU=dae,OU=people,O=u.s. government,C=us
   - subject = twl-server-generic2,OU=DAE,OU=DIA,OU=twl-server-generic2,O=U.S. Government,C=US
   - FAIL! Unable to find info that this is from tester10!!!!
   - Operation fails because twl-server-generic2 doesnt have access to tester10 stuff in odrive

2. When connecting as twl-server-generic2
   - Headers sent from client: 
     - user_dn: cn=test tester10,ou=people,ou=dae,ou=chimera,o=u.s. government,c=us
     - external_sys_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
     - ssl_client_s_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
   - AuthInfo has chain of
     - CN=twl-server-generic2,OU=DAE,OU=DIA,OU=twl-server-generic2,O=U.S. Government,C=US
     - CN=InterCA,OU=chimera,OU=dae,OU=people,O=u.s. government,C=us
   - In the whitelist, so able to impersonate as user_dn
   - FAIL! .. no validation of the original caller. Effectively same as non TLS (see below)

# Results Using TLS for only GRPC

settings.toml...
- grpc_use_tls = true
- gateway_use_tls = false
- use_gateway_proxy = true
- server_cert_path/key .. server.cert/key.pem for twl-server-generic2

1. When connecting as anonymous and HTTP
   - Headers sent from client: 
     - user_dn: cn=test tester10,ou=people,ou=dae,ou=chimera,o=u.s. government,c=us
     - external_sys_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
     - ssl_client_s_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
   - AuthInfo has chain of
     - CN=twl-server-generic2,OU=DAE,OU=DIA,OU=twl-server-generic2,O=U.S. Government,C=US
     - CN=InterCA,OU=chimera,OU=dae,OU=people,O=u.s. government,C=us
   - Able to impersonate as user_dn
   - FAIL! no validation here

# Results Using TLS for only Gateway

settings.toml...
- grpc_use_tls = false
- gateway_use_tls = true
- use_gateway_proxy = true
- server_cert_path/key .. server.cert/key.pem for twl-server-generic2

1. When connecting as tester10,
   - Headers sent from client: none 
   - Routed through as HTTP, and no headers passed
   - FAIL! invalid connection. required headers are missing

2. When connecting as twl-server-generic2
   - Headers sent from client: 
     - user_dn: cn=test tester10,ou=people,ou=dae,ou=chimera,o=u.s. government,c=us
     - external_sys_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
     - ssl_client_s_dn: cn=twl-server-generic2,ou=dae,ou=dia,ou=twl-server-generic2,o=u.s. government,c=us
   - AuthInfo has nothing!
   - Able to impersonate since its passed in the header..
   - FAIL! there is no peer certificate to verify!

### PR DESCRIPTION
**Pull request by dougfort**
closes #166 

sample method:
```go
ackage methods

import (
	"time"

	gometrics "github.com/armon/go-metrics"

	"github.com/deciphernow/gm-fabric-go/tlsutil"

	pb "testdir/test_service/protobuf"
)

// HelloStream says "hello" repeatedly in a stream
func (s *serverData) HelloStream(req *pb.HelloStreamRequest, stream pb.TestService_HelloStreamServer) error {

	defer gometrics.MeasureSince(
		[]string{
			"test_service", // service name
			"HelloStream",
			"elapsed",
		},
		time.Now(),
	)

	s.Logger.Debug().Str("name", tlsutil.GetDNFromContext(stream.Context())).Msg("")

	for i := int32(0); i < req.Count; i++ {
		stream.Send(&pb.HelloResponse{Text: "pong"})
	}

	return nil
}
```

output
```
2018-02-15T13:02:05-05:00 |DEBU| loading TLS config ca_cert_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/intermediate.crt server_cert_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/localhost.crt server_key_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/localhost.key service=test_service
2018-02-15T13:02:05-05:00 |DEBU| loading TLS config ca_cert_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/intermediate.crt server_cert_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/localhost.crt server_key_path=/home/dougfort/fabric_test_dir/src/testdir/test_service/localhost.key service=test_service
2018-02-15T13:02:05-05:00 |DEBU| creating listener host= port=10000 service=test_service
2018-02-15T13:02:05-05:00 |DEBU| starting metrics server host= port=10001 service=test_service
2018-02-15T13:02:05-05:00 |DEBU| starting grpc server service=test_service
2018-02-15T13:02:05-05:00 |DEBU| starting gateway proxy service=test_service
2018-02-15T13:02:05-05:00 |DEBU| starting gateway proxy server host= port=8080 service=test_service
2018-02-15T13:02:13-05:00 |DEBU| <nil> name="CN=localhost,L=Alexandria,ST=Virginia,OU=Engineering,O=Decipher Technology Studios,C=US"
2018-02-15T13:02:13-05:00 |DEBU| <nil> name="CN=localhost,L=Alexandria,ST=Virginia,OU=Engineering,O=Decipher Technology Studios,C=US"
2018-02-15T13:02:19-05:00 |INFO| shutting down service=test_service signal=terminated

```

### CODE REVIEW

File: tlsutil/tlsclient.go [hunk 1]
```diff
@@ -55,7 +59,7 @@ func WithClientAuth(auth tls.ClientAuthType) Opt {
 func NewTLSConfig(trustPath, certPath, keyPath string, opts ...Opt) (*tls.Config, error) {
 	trustBytes, err := ioutil.ReadFile(trustPath)
 	if err != nil {
-		return nil, fmt.Errorf("error parsing CA trust %s: %v", trustPath, err)
+		return nil, fmt.Errorf("error parsinGetDNFromContextg CA trust %s: %v", trustPath, err)
```
Comment (lucasmoten) [inline tlsutil/tlsclient.go]: Should correct this spelling mishap (search/replace?) in this error

**Comment by lucasmoten** [comment]
Where/How do you get this function signature? What does the original proto file look like..
```
func (s *serverData) HelloStream(req *pb.HelloStreamRequest, stream pb.TestService_HelloStreamServer) error {
```
I dont have anything like that resulting from my proto files.

File: tlsutil/tlsclient.go (hunk 1 above)
Comment (dougfort) [reply tlsutil/tlsclient.go]: oops

**Comment by dougfort** [comment]
That signature comes from https://github.com/DecipherNow/gm-fabric-go/blob/166-identify-peer/cmd/fabric/test_fabric.sh

That's my internal regression test.

The proto file looks like
```protobuf
syntax = "proto3";

package protobuf;

import "google/api/annotations.proto";

// Interface exported by the server.
service TestService {
    // HelloProxy says 'hello' in a form that is handled by the gateway proxy
	rpc HelloProxy(HelloRequest) returns (HelloResponse) {
		option (google.api.http) = {
            get: "/acme/services/hello"
        };
    }

    // HelloStream returns multiple replies
    rpc HelloStream(HelloStreamRequest) returns (stream HelloResponse) {}

}
message HelloRequest {
    string hello_text = 1;
}

message HelloStreamRequest {
    string hello_text = 1;
    int32 count = 2;
}

message HelloResponse {
	string text = 1;
}
```

The generated interface looks like:
```go
// Server API for TestService service

type TestServiceServer interface {
	// HelloProxy says 'hello' in a form that is handled by the gateway proxy
	HelloProxy(context.Context, *HelloRequest) (*HelloResponse, error)
	// HelloStream returns multiple replies
	HelloStream(*HelloStreamRequest, TestService_HelloStreamServer) error
}

// some other stuff

type TestService_HelloStreamServer interface {
	Send(*HelloResponse) error
	grpc.ServerStream
}
```

**Comment by lucasmoten** [comment]
These two methods are nice additions
- GetClientCertificatesFromContext
- GetDNFromContext

However, this wont technically close #166 since they dont address the gateway to grpc conversion where the issue arises.  In the gateway to grpc translation, something needs to pickup the peer info and pass it along into grpc. That's why i did it with an option as follows

```
// TLSPeerOption sets a metadata value to expose the distinguished name of original connecting peer
func TLSPeerOption() runtime.ServeMuxOption {
	f := func(ctx oldcontext.Context, req *http.Request) metadata.MD {
		metaMap := make(map[string]string)

		if len(req.TLS.PeerCertificates) > 0 {
			metaMap["tls_peer_certificate"] = tlsutil.GetDistinguishedName(req.TLS.PeerCertificates[0])
		}
		return metadata.New(metaMap)
	}
	return runtime.WithMetadata(f)
}
```

**Comment by dougfort** [comment]
Why pass it to the gRPC method when the method can get the information for itself?

**Comment by dougfort** [comment]
Oh, I think I see what you mean. That the gateway isn't passing the peer information correctly. 

Can you give me enough information to create a test that fails? 

This works OK: finds the cert in the gRPC methd
```bash
$ ~/fabric_test_dir/bin/test_service_http_client --uri="https://127.0.0.1:8080/acme/services/hello?hello_text=ping" --test-cert-dir=. > "/home/dougfort/fabric_test_dir/http_client.log" 2>&1
```

**Comment by alecholmez** [comment]
@dougfort since we are rolling this into `tlsutil` you might want to delete the existing dn from context code in the `rpcutil` package. Don't want redundancy

**Comment by dougfort** [comment]
@lucasmoten says this change does not satisfy #166. I need a failing test case to continue with this.

**Comment by lucasmoten** [comment]
I wont have time to work on this since i've already solved teh problem in cogito and wrote up notes in #166 describing what happens when different certificates are configured.Go ahead an rebase and merge, and I'll just file a bug report in the future when I need to build another service that needs to know who the user is

**Pull request merged**